### PR TITLE
Add CL4RRedeemer

### PR DESCRIPTION
**Pull request by ben-kaufman**

### CODE REVIEW

File: contracts/schemes/CL4RRedeemer.sol [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    uint256 public startTime; // the time (in secs since epoch) that locking can start (is enable)
+    uint256 public redeemEnableTime;
+    uint256 public maxLockingBatches;
+    uint256 public batchTime; // the length of a batch, in seconds
+    IERC20 public token; // the token to be locked
+    uint256 public lockCounter; // Total number of locks
+    uint256 public totalLockedLeft; // the amount of reputation  that is still left to distribute
+    uint256 public repRewardConstA;
+    uint256 public repRewardConstB;
+    uint256 public batchesIndexCap;
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:46]: unused vars

File: contracts/schemes/CL4RRedeemer.sol [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        uint256 period;
+    }
+
+    // A mapping from lockers addresses to their locks
+    mapping(address => mapping(uint256=>Lock)) public lockers;
+    // A mapping from batch index to batch
+    mapping(uint256 => Batch) public batches;
+
+    uint256 public reputationRewardLeft; // the amount of reputation that is still left to distribute
+    uint256 public startTime; // the time (in secs since epoch) that locking can start (is enable)
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:37]: can these vars be read as local vars at the redeem function ?

File: contracts/schemes/CL4RRedeemer.sol [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    struct Lock {
+        uint256 amount;
+        uint256 lockingTime;
+        uint256 period;
+    }
+
+    // A mapping from lockers addresses to their locks
+    mapping(address => mapping(uint256=>Lock)) public lockers;
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:32]: unused ?

File: contracts/schemes/CL4RRedeemer.sol [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+     * @dev redeem reputation function
+     * @param _beneficiary the beneficiary to redeem.
+     * @param _lockingId the lockingId to redeem from.
+     * @return reputation reputation rewarded
+     */
+    function redeem(address _beneficiary, uint256 _lockingId) public returns(uint256 reputation) {
+
+        // solhint-disable-next-line not-rely-on-time
+        require(now > redeemEnableTime, "now > redeemEnableTime");
+        Lock storage locker = lockers[_beneficiary][_lockingId];
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:84]: `Lock memory locker;`

File: contracts/schemes/CL4RRedeemer.sol (hunk 2 above)
Comment (ben-kaufman) [reply contracts/schemes/CL4RRedeemer.sol:37]: Yes, but that'd be wasteful to read them every call as they are constants.

Comment (orenyodfat) [reply contracts/schemes/CL4RRedeemer.sol:35]: in the cost of high deployment gas

Comment (ben-kaufman) [reply contracts/schemes/CL4RRedeemer.sol:35]: But then it's lower cost for using the contract. It's one time higher payment for multiple times lower payments tradeoff. I can change that but I think like this is better for gas saving.

File: contracts/schemes/CL4RRedeemer.sol [hunk 5]
```diff
@@ -0,0 +1,141 @@
+pragma solidity ^0.6.12;
+// SPDX-License-Identifier: GPL-3.0
+
+import "./ContinuousLocking4Reputation.sol";
+
+/**
+ * @title A scheme for continuous locking ERC20 Token for reputation
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:7]: please add a short description of this contract .
why it is needed and how it is suppose to work.

File: contracts/schemes/CL4RRedeemer.sol [hunk 6]
```diff
@@ -0,0 +1,146 @@
+pragma solidity ^0.6.12;
+// SPDX-License-Identifier: GPL-3.0
+
+import "./ContinuousLocking4Reputation.sol";
+
+/**
+ * @title A scheme for redeeming a ContinuousLocking4Reputation contract used in a different DAO.
+ * This is used to migrate from an older DAO to a new one while the older DAO has an active ContinuousLocking4Reputation
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:8]: `This is used` - > `This contract can be use `

File: contracts/schemes/CL4RRedeemer.sol [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+ * Using it should be done by initiaalizing this contract s a scheme in the new DAO,
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:9]: `initiaalizing` -> `initializing`

`contract s a scheme` -> contract as a scheme

File: contracts/schemes/CL4RRedeemer.sol [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+ * with the old ContinuousLocking4Reputation as an init parameter.
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:10]: as an init parameter (_ct4r)

File: contracts/schemes/CL4RRedeemer.sol [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+ * Then the old ContinuousLocking4Reputation should be unregistered as a scheme from the older DAO so
```
Comment (orenyodfat) [inline contracts/schemes/CL4RRedeemer.sol:11]: Note : the old ContinuousLocking4Reputation should be unregistered as a scheme from the older  ....

**Comment by ben-kaufman** [comment]
Fixed the description @orenyodfat

**Review by orenyodfat — APPROVED** [review]

**Pull request merged**